scripts/train/train.py

```diff
@@ -106,21 +182,54 @@ def validate_config(cfg: DictConfig):
                                            'mptmlp') in ffns_with_megablocks:
         moe_world_size = cfg.model.get('ffn_config',
                                        {}).get('moe_world_size', 1)
-        use_orig_params = cfg.get('fsdp_config',
-                                  {}).get('use_orig_params', True)
+        use_orig_params = cfg.fsdp_config.get(
+            'use_orig_params', True) if cfg.fsdp_config is not None else True
         if moe_world_size > 1 and not use_orig_params:
             raise ValueError(
                 f'MoEs with expert parallelism (moe_world_size {moe_world_size} > 1) require `use_orig_params=True`.'
             )
 
 
 def main(cfg: DictConfig) -> Trainer:
-    # Run user provided code if specified
-    code_paths = pop_config(cfg,
-                            'code_paths',
-                            must_exist=False,
-                            default_value=[],
-                            convert=True)
+    # Resolve all interpolation variables as early as possible
+    om.resolve(cfg)
+
+    # structured config does not support unions of containers, so separate single and plural containers
+    if (loader := cfg.get('eval_loader', None)) is not None:
+        if isinstance(loader, ListConfig):
+            cfg['eval_loaders'] = list(cfg.pop('eval_loader'))
+    if (tasks := cfg.get('icl_tasks', None)) is not None:
+        if isinstance(tasks, str):
+            cfg['icl_tasks_str'] = cfg.pop('icl_tasks')
+    if (gauntlet := cfg.get('eval_gauntlet', None)) is not None:
+        if isinstance(gauntlet, str):
+            cfg['eval_gauntlet_str'] = cfg.pop('eval_gauntlet')
+
+    arg_config_keys = set(cfg.keys())
+    extraneous_keys = set.difference(arg_config_keys, TRAIN_CONFIG_KEYS)
+
+    if 'variables' not in cfg:
+        cfg['variables'] = {}
+
+    for key in extraneous_keys:
+        warnings.warn(
+            f'Unused parameter {key} found in cfg. Please check your yaml to ensure this parameter is necessary. Interpreting {key} as a variable for logging purposes.'
+        )
+        # TODO (milo): delete the below line once we deprecate variables at the top level.
+        cfg['variables'][key] = cfg.pop(key)
+
+    # Create copy of config for logging
+    logged_cfg: DictConfig = copy.deepcopy(cfg)
+
+    # Get global and device batch size information from distributed/single node setting
+    cfg = update_batch_size_info(cfg)
+    logged_cfg.update(cfg, merge=True)
+
+    scfg: TrainConfig = OmegaConf.structured(
```

Comment:
nit: `cfg` and `scfg` are pretty similar, worried about typos here. Could we call the input `unstructured_cfg` and the TrainConfig `cfg`? Or something to that effect. Just want to make sure their names are a little more obviously different